daemon/archive.go

```diff
@@ -152,7 +153,28 @@ func (daemon *Daemon) containerArchivePath(container *container.Container, path
 	// container ID.
 	opts := archive.TarResourceRebaseOpts(resolvedPath, filepath.Base(absPath))
 
-	data, err := chrootarchive.Tar(resolvedPath, opts, filepath.Dir(absPath))
+	// Get the directory where the file lives in the container.
+	ctrDir := filepath.Join(container.BaseFS, filepath.Dir(absPath))
+
+	// Evaluate any symlink that occurs and point to the actual dir.
+	// EvalSymlinks will look for the target on the host and will fail
+	// if not found.  Ignore errors and assume it's OK in the container.
+	resolvedCtrDir, _ := filepath.EvalSymlinks(ctrDir)
+
+	fileInfo, err := os.Lstat(ctrDir)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error retrieving fileinfo for %q", ctrDir)
+	}
+	// If resolvedCtrDir was a symlink, it will have the resolved directory from
+	// the containers root.  i.e. `ln -s /tmp /test` in the container and
+	// resolvedCtrDir would have `/tmp` in it at this point.  Add it back
+	// to the container.Basefs to get the full container target directory
+	// on the host.
+	if fileInfo.Mode()&os.ModeSymlink != 0 {
+		resolvedCtrDir = filepath.Join(container.BaseFS, resolvedCtrDir)
```

Comment:
should we ensure that after the Join the path is still under `container.BaseFS`?

Could happen that `resolvedCtrDir` has `..` components, so it can potentially point outside of the basefs once we do the join?

Reply:
I played with it tonight, and excellent spot.  I think I've a solution for that, testing now.

Reply:
Actually, I played more and add a symlink in the container like:  `ln -s /tmp ../../../../../../testy` and that just resolved to the root directory of the container. However I did come up with a quick check that will make it completely safe, so I've added it.